### PR TITLE
API endpoint for manually relaying warp message

### PR DESCRIPTION
**Pull request by geoff-vball**
## Why this should be merged
Closes https://github.com/ava-labs/awm-relayer/issues/314

## How this works
Creates a global `MessageCoordinator` that contains maps to Source Clients, Application Relayers, and Message Handler Factories.

TODOs:
- Make block number optional (implement iterative look-back for the message)
- Flag to enable/disable the API?
- Documentation around potential DOS vectors of the API should it be open to the public networking wise?
- Should any authentication be required? Probably makes sense to recommend it is only an RPC available locally.
- Currently we have no way to distinguish between the transaction being successfully sent and `ShouldSendMessage()` returning false.

## How this was tested
E2E Test

## How is this documented
TODO

### CODE REVIEW

File: relayer/relay_message_api_handler.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"encoding/json"
+	"math/big"
+	"net/http"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const RelayMessageApiPath = "/relay-message"
```
Comment (cam-schultz) [inline relayer/relay_message_api_handler.go:12]: I prefer single name paths. Here, I think `/relay` would suffice.

File: relayer/relay_message_api_handler.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+type RelayMessageRequest struct {
```
Comment (cam-schultz) [inline relayer/relay_message_api_handler.go:14]: Let's document that all of these params are required.

File: relayer/relay_message_api_handler.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	// cb58 encoding of the blockchain ID
+	BlockchainID string `json:"blockchain-id"`
+	// Hex encoding of the warp message ID
+	MessageID string `json:"message-id"`
+	// Integer representation of the block number
+	BlockNum string `json:"block-num"`
```
Comment (cam-schultz) [inline relayer/relay_message_api_handler.go:20]: Off-chain Warp messages do not have a block number associated with them. We should provide a route for fully specifying a Warp message so that we don't have to fetch from the source chain at a specific height.

I'm thinking we have `/relay` that accepts a Warp message ID as well as a height, and a /relay/message` that accepts the full unsigned message bytes, but no height.

File: relayer/relay_message_api_handler.go
Comment (cam-schultz) [inline relayer/relay_message_api_handler.go:1]: Let's document this endpoint in the top-level README

File: relayer/relay_message_api_handler.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+		err := json.NewDecoder(r.Body).Decode(&req)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		blockchainID, err := ids.FromString(req.BlockchainID)
+		if err != nil {
+			http.Error(w, "invalid blockchainID"+err.Error(), http.StatusBadRequest)
```
Comment (cam-schultz) [inline relayer/relay_message_api_handler.go:35]:
```suggestion
			http.Error(w, "invalid blockchainID: "+err.Error(), http.StatusBadRequest)
```

**Review by cam-schultz — COMMENTED** [review]
I left a handful of comments and asks, and I'd like to think a bit more about the new `MessageCoordinator` type. But overall, the changes make sense and I'm a big fan of the trend of decoupling all of the components and coordinating them in `main`.

File: main/main.go [hunk 5]
```diff
@@ -162,55 +167,72 @@ func main() {
...
+	relayer.SetMessageCoordinator(logger, messageHandlerFactories, applicationRelayers, sourceClients)
+
+	// Initialize the API after the message coordinator is set
+	http.Handle("/health", health.NewHandler(checker))
+	http.Handle(relayer.RelayMessageApiPath, relayer.RelayMessageAPIHandler())
+
+	// start the health check server
+	go func() {
+		log.Fatalln(http.ListenAndServe(fmt.Sprintf(":%d", cfg.APIPort), nil))
+	}()
```
[the remark this answers is not in the document]
Comment (geoff-vball) [reply main/main.go:227]: I'm not sure exactly what you mean here. Do you mean instantiating a new `ServeMux` instead of using the default one for everything here? Or encapsulating all this logic in a separate package?

File: main/main.go [hunk 6]
```diff
@@ -102,18 +106,22 @@ func main() {
...
-		logger.Error(
-			"Failed to create destination clients",
-			zap.Error(err),
-		)
+		logger.Fatal("Failed to create destination clients", zap.Error(err))
+		panic(err)
+	}
+
+	// Initialize all source clients
+	logger.Info("Initializing destination clients")
```
Comment (minghinmatthewlam) [inline main/main.go:114]:
```suggestion
	logger.Info("Initializing source clients")
```

File: main/main.go [hunk 7]
```diff
@@ -162,55 +167,72 @@ func main() {
...
 		panic(err)
 	}
 
 	// Initialize the global write ticker
 	ticker := utils.NewTicker(cfg.DBWriteIntervalSeconds)
 	go ticker.Run()
 
+	messageHandlerFactories, err := createMessageHandlerFactories(logger, &cfg)
+	if err != nil {
+		logger.Fatal("Failed to create Message Handler Factories", zap.Error(err))
```
Comment (minghinmatthewlam) [inline main/main.go:198]:
```suggestion
		logger.Fatal("Failed to create message handler factories", zap.Error(err))
```

File: main/main.go [hunk 8]
```diff
@@ -227,136 +249,161 @@ func main() {
...
+			case config.OFF_CHAIN_REGISTRY:
+				m, err = offchainregistry.NewMessageHandlerFactory(
+					logger,
+					cfg,
+				)
+			default:
+				m, err = nil, fmt.Errorf("invalid message format %s", format)
+			}
+			if err != nil {
+				logger.Error("Failed to create message manager", zap.Error(err))
```
Comment (minghinmatthewlam) [inline main/main.go:310]:
```suggestion
				logger.Error("Failed to create message handler factory", zap.Error(err))
```

File: main/main.go [hunk 9]
```diff
@@ -227,136 +249,161 @@ func main() {
...
-		applicationRelayers,
-		minHeight,
-		ethClient,
-	)
-	if err != nil {
-		return fmt.Errorf("failed to create listener instance: %w", err)
+) (map[ids.ID]map[common.Address]messages.MessageHandlerFactory, error) {
+	messageHandlerFactories := make(map[ids.ID]map[common.Address]messages.MessageHandlerFactory)
+	for _, sourceBlockchain := range globalConfig.SourceBlockchains {
+		messageHandlerFactoriesForSource := make(map[common.Address]messages.MessageHandlerFactory)
```
Comment (minghinmatthewlam) [inline main/main.go:285]: nit: can we rename to `messageProtocolHandlerFactory` or simiilar

File: main/main.go [hunk 10]
```diff
@@ -162,55 +167,72 @@ func main() {
...
+		db,
+		ticker,
+		network,
+		messageCreator,
+		&cfg,
+		sourceClients,
+		destinationClients,
+	)
+	if err != nil {
+		logger.Fatal("Failed to create Application Relayers", zap.Error(err))
```
Comment (minghinmatthewlam) [inline main/main.go:215]:
```suggestion
		logger.Fatal("Failed to create application relayers", zap.Error(err))
``

File: relayer/message_coordinator.go [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/messages"
+	relayerTypes "github.com/ava-labs/awm-relayer/types"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+var globalMessageCoordinator *MessageCoordinator
```
Comment (minghinmatthewlam) [inline relayer/message_coordinator.go:22]: I think we've generally avoided global vars, @cam-schultz is that correct? My personal preference has been to avoid them.

If we do decide to go with them though, should we have some checks that these are singleton instances?

File: main/main.go [hunk 12]
```diff
@@ -132,10 +140,7 @@ func main() {
 		&cfg,
 	)
 	if err != nil {
-		logger.Error(
-			"Failed to create app request network",
-			zap.Error(err),
-		)
+		logger.Fatal("Failed to create app request network", zap.Error(err))
 		panic(err)
 	}
```
Comment (minghinmatthewlam) [inline main/main.go:146]: not part of this PR, but what hapepns when the function callback returns an error, I'm not sure where it gets handled, and how it's handled.

Comment (geoff-vball) [reply main/main.go:146]: Which function callback do you mean?

File: main/main.go (hunk 9 above)
Comment (geoff-vball) [reply main/main.go:285]: Rename which?

File: relayer/relay_message_api_handler.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	"math/big"
+	"net/http"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const RelayMessageApiPath = "/relay"
+
+type RelayMessageRequest struct {
```
Comment (minghinmatthewlam) [inline relayer/relay_message_api_handler.go:14]: should we have a `RelayMessageResponse` type as well?

File: relayer/message_coordinator.go (hunk 11 above)
Comment (geoff-vball) [reply relayer/message_coordinator.go:22]: Michael, Cam, and I landed on having a global object for these shared resources. With the current design, `ApplicationRelayer`s are inherently global, and we need some way to access them. We can't access the right one without the `MessageHandlerFactories`.

We _could_ make this is singleton, but I don't really see a benefit of enforcing only one instantiation of it (though I don't think we _should_ instantiate more than 1)

File: relayer/relay_message_api_handler.go (hunk 3 above)
Comment (geoff-vball) [reply relayer/relay_message_api_handler.go:20]: In order to process a message in the relayer we need: `NetworkID`, `SourceAddress`, `SourceChainID`, and `Payload`. Do we want to specify all of these for this endpoint?

Comment (geoff-vball) [reply relayer/relay_message_api_handler.go:20]: The idea is to make the block number optional later, but it requires implementing an iterative lookback for on-chain messages.

File: main/main.go (hunk 9 above)
Comment (minghinmatthewlam) [reply main/main.go:259]: `messageHandlerFactoriesForSource`

File: main/main.go (hunk 12 above)
Comment (minghinmatthewlam) [reply main/main.go:143]: the health checker's check function https://github.com/ava-labs/awm-relayer/pull/327/files/6dc2441bb03dc49a9e1be1f5c768c913dda73254#diff-327181d0a8c5e6b164561d7910f4eeffd41442d55b2a2788fda2aa2692f17ec0R163

File: relayer/message_coordinator.go (hunk 11 above)
Comment (minghinmatthewlam) [reply relayer/message_coordinator.go:22]: was just thinking to have a builtin check that makes sure `globalMessageCoordinator` isn't overrode with another instance

File: main/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cam-schultz) [reply main/main.go]: Sorry, my original comment was pretty unspecific. Now that we have two distinct endpoints, it might make sense to put them and their handlers into their own package, rather than the current organization with one handler in package `main`, and another in package `relayer`. I don't feel super strongly though, especially if it is a significant rework or breaks the existing dependency graph.

**Review by bernard-avalabs — COMMENTED** [review]
Generally makes sense to me. I've left a few minor comments.

File: relayer/relay_message_api_handler.go (hunk 3 above)
Comment (cam-schultz) [reply relayer/relay_message_api_handler.go:26]: I see two options:
1. Provide a separate endpoint that fully specifies the off-chain Warp message, including the fields you mentioned
2. Implement control flow to fetch the off-chain Warp message by it's ID using the Warp API. The downside here is that the public APIs disable the Warp API.

File: relayer/message_coordinator.go (hunk 11 above)
Comment (cam-schultz) [reply relayer/message_coordinator.go:22]: It's a semantic difference, but we have a bunch of variables that live at `package main` scope and are passed around, as opposed to static global vars that live at application scope. We confusingly use the the term "global" for the former. 

They're functionally the same, but in order to remain consistent, we should return the `MessageCoordinator` instance to `main` and leave it to `main` to ensure that only one instance exists. We can then pass the `MessageCoordinator` reference to all of the downstream objects that interact with it, similar to how we treat the `RelayerDatabase` and `AppRequestNetwork` global singletons.

File: main/main.go (hunk 12 above)
Comment (geoff-vball) [reply main/main.go:143]: It returns a `503 (Service Unavailable)` HTTP response. https://github.com/alexliesenfeld/health/blob/main/examples/showcase/main.go

File: main/main.go [hunk 14]
```diff
@@ -227,136 +249,161 @@ func main() {
...
-					zap.String("blockchainID", sourceBlockchain.BlockchainID),
-					zap.Error(err),
-				)
-				return err
-			}
-			logger.Info(
-				"Created application relayers",
-				zap.String("blockchainID", sourceBlockchain.BlockchainID),
-			)
+			return relayer.ProcessManualWarpMessages(manualWarpMessages[sourceBlockchain.GetBlockchainID()])
```
[the remark this answers is not in the document]
Comment (geoff-vball) [reply main/main.go]: Sorry, I already removed the manual warp messages from the config. Do you think we should keep them in there for a few releases with a deprecation note? I can't imagine many people were using them already given the difficulty in having to add them to the nodes as well.

[the remark this answers is not in the document]
Comment (cam-schultz) [reply main/main.go]: Yeah, we should deprecate features gracefully at this point, even though I agree that likely few people are using this particular feature. Apologies for not making that clear up front

[the remark this answers is not in the document]
Comment (geoff-vball) [reply main/main.go]: So a couple thoughts here. We don't initialize the logger for the application until after we finish parsing the config. So it would be a little hard to emit logs from here. Returning an error and failing validation seems a little too extreme here, is there some middle ground?

Do we still want to process the manual warp messages if they're included in the config? It should be easy enough to point the users to the new API.

File: main/main.go [hunk 15]
```diff
@@ -191,180 +159,209 @@ func main() {
...
+		sourceClients,
+		destinationClients,
+	)
+	if err != nil {
+		logger.Fatal("Failed to create application relayers", zap.Error(err))
+		panic(err)
 	}
+	messageCoordinator := relayer.NewMessageCoordinator(logger, messageHandlerFactories, applicationRelayers, sourceClients)
+
+	// Each goroutine will have an atomic bool that it can set to false if it ever disconnects from its subscription.
```
Comment (cam-schultz) [inline main/main.go:203]:
```suggestion
	// Each Listener goroutine will have an atomic bool that it can set to false to indicate an unrecoverable error
```

File: main/main.go [hunk 16]
```diff
@@ -191,180 +159,209 @@ func main() {
...
-		minHeight,
-		ethClient,
-	)
-	if err != nil {
-		return fmt.Errorf("failed to create listener instance: %w", err)
+) (map[ids.ID]map[common.Address]messages.MessageHandlerFactory, error) {
+	messageHandlerFactories := make(map[ids.ID]map[common.Address]messages.MessageHandlerFactory)
+	for _, sourceBlockchain := range globalConfig.SourceBlockchains {
+		messageHandlerFactoriesForSource := make(map[common.Address]messages.MessageHandlerFactory)
+		// Create message managers for each supported message protocol
```
Comment (cam-schultz) [inline main/main.go:244]:
```suggestion
		// Create message handler factories for each supported message protocol
```

File: messages/message_handler.go [hunk 17]
```diff
@@ -27,7 +27,7 @@ type MessageHandler interface {
 
 	// SendMessage sends the signed message to the destination chain. The payload parsed according to
 	// the VM rules is also passed in, since MessageManager does not assume any particular VM
```
Comment (cam-schultz) [inline messages/message_handler.go:29]: Can you add to the interface comment what the returned common.Hash is intended to represent?

File: relayer/application_relayer.go [hunk 18]
```diff
@@ -182,19 +184,17 @@ func (r *ApplicationRelayer) ProcessHeight(height uint64, handlers []messages.Me
 }
 
 // Relays a message to the destination chain. Does not checkpoint the height.
```
Comment (cam-schultz) [inline relayer/application_relayer.go:186]: Can you add to the funciton comment what the new return value represents?

File: tests/basic_relay.go [hunk 19]
```diff
@@ -114,8 +114,8 @@ func BasicRelay(network interfaces.LocalNetwork) {
 	relayerIDA := database.CalculateRelayerID(subnetAInfo.BlockchainID, subnetBInfo.BlockchainID, database.AllAllowedAddress, database.AllAllowedAddress)
 	relayerIDB := database.CalculateRelayerID(subnetBInfo.BlockchainID, subnetAInfo.BlockchainID, database.AllAllowedAddress, database.AllAllowedAddress)
 	// Modify the JSON database to force the relayer to re-process old blocks
-	jsonDB.Put(relayerIDA, database.LatestProcessedBlockKey, []byte("0"))
-	jsonDB.Put(relayerIDB, database.LatestProcessedBlockKey, []byte("0"))
+	_ = jsonDB.Put(relayerIDA, database.LatestProcessedBlockKey, []byte("0"))
+	_ = jsonDB.Put(relayerIDB, database.LatestProcessedBlockKey, []byte("0"))
```
Comment (cam-schultz) [inline tests/basic_relay.go:118]: Let's check the returned errors here, else leave it as is.

File: tests/relay_message_api.go
Comment (cam-schultz) [inline tests/relay_message_api.go:1]: Do we check that the message was successfully relayed in this test? If not, let's add that in.

File: tests/utils/utils.go [hunk 20]
```diff
@@ -204,6 +204,7 @@ func CreateDefaultRelayerConfig(
 		MetricsPort:            9090,
 		SourceBlockchains:      sources,
 		DestinationBlockchains: destinations,
+		APIPort:                1234,
```
Comment (cam-schultz) [inline tests/utils/utils.go:207]: Any reason not to use the default port 8080? That's what will be used if this field is not explicitly set.

File: relayer/message_coordinator.go [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+			zap.Error(err),
+			zap.String("warpMessageID", warpMessage.UnsignedMessage.ID().String()),
+		)
+		return common.Hash{}, err
+	}
+
+	return appRelayer.ProcessMessage(handler)
+}
+
+func (mc *MessageCoordinator) ProcessMessage(blockchainID ids.ID, messageID common.Hash, blockNum *big.Int) (common.Hash, error) {
```
Comment (cam-schultz) [inline relayer/message_coordinator.go:187]: I'd suggest renaming this to `ProcessMessageID` to distinguish it from `ApplicationRelayer.ProcessMessage`, which operates on the full message, and doesn't fetch it from the chain.

**Review by cam-schultz — COMMENTED** [review]
Just a handful of minor comments, otherwise LGTM.

Let's also make sure to document all API routes (new and existing) in the README, and confirm with stakeholders that removing the manual warp message config won't break any existing functionaliity.

**Review by michaelkaplan13 — COMMENTED** [review]
Could you add an example relay message request & responses to the README?

File: api/relay_message.go [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/ava-labs/awm-relayer/relayer"
+	"github.com/ava-labs/awm-relayer/types"
+	relayerTypes "github.com/ava-labs/awm-relayer/types"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	RelayApiPath        = "/relay"
```
Comment (michaelkaplan13) [inline api/relay_message.go:17]:
```suggestion
	RelayAPIPath        = "/relay"
```

File: types/types.go [hunk 23]
```diff
@@ -107,3 +108,20 @@ func UnpackWarpMessage(unsignedMsgBytes []byte) (*avalancheWarp.UnsignedMessage,
...
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch logs: %w", err)
+	}
+	if len(logs) != 1 {
+		return nil, fmt.Errorf("found more than 1 log: %d", len(logs))
+	}
+
+	return NewWarpMessageInfo(logs[0])
+}
```
Comment (michaelkaplan13) [inline types/types.go:127]: This feels a little odd to be in the `types` package. I think either `utils` or `relayer` would probably be a better fit for it.

File: tests/utils/utils.go (hunk 20 above)
Comment (geoff-vball) [reply tests/utils/utils.go:207]: I just needed to use something. The defaults don't get set in the tests. I've changed it to 8080

File: tests/basic_relay.go (hunk 19 above)
Comment (geoff-vball) [reply tests/basic_relay.go]: This sent me down a small rabbit hole... These were actually returning an error because the config wasn't validated, the `SupportedDestination` fields weren't populated, so `database.NewJSONFileStorage` wasn't populated with the correct relayer IDs.

I think it might make sense to add a `validated` field to the config struct, so that certain functions can check if the config has been validated before operating on it.

File: api/relay_message.go [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+	MessageID string `json:"message-id"`
+	// Required. Integer representation of the block number
+	BlockNum string `json:"block-num"`
+}
+
+// Defines a manual warp message to be sent from the relayer through the API.
+type ManualWarpMessageRequest struct {
+	UnsignedMessageBytes []byte
+	SourceAddress        common.Address
+}
```
[the remark this answers is not in the document]
Comment (geoff-vball) [reply api/relay_message.go:34]: This can be used for off-chain messages as well as on-chain messages if you want to provide the message manually.

File: relayer/message_coordinator.go [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+	if !ok {
+		mc.logger.Error("Source client not found", zap.String("blockchainID", blockchainID.String()))
+		return common.Hash{}, fmt.Errorf("source client not set for blockchain: %s", blockchainID.String())
+	}
+
+	warpMessage, err := relayerTypes.FetchWarpMessageFromID(ethClient, messageID, blockNum)
+	if err != nil {
+		mc.logger.Error("Failed to fetch warp from blockchain", zap.String("blockchainID", blockchainID.String()), zap.Error(err))
+		return common.Hash{}, fmt.Errorf("could not fetch warp message from ID: %w", err)
+	}
```
Comment (cam-schultz) [inline relayer/message_coordinator.go]: To simplify the `MessageCoorinator` exported interface, how about we lift this logic out to the caller, and combine `ProcessMessage` and `ProcessManualWarpMessage`?

Comment (geoff-vball) [reply relayer/message_coordinator.go]: Good call :)

File: tests/basic_relay.go (hunk 19 above)
Comment (cam-schultz) [reply tests/basic_relay.go]: Rather than adding a field that's unused in the main application, I'd prefer to instead make sure `Config.Validate` is idempotent, and call it here in the test. On inspection, I think it should already be idempotent.

File: relayer/message_coordinator.go (hunk 25 above)
Comment (cam-schultz) [reply relayer/message_coordinator.go]: Yeah, we'd need to decide if the interface should accept only Warp Messages, or also IDs. Only accepting Warp Messages would yield a smaller interface, but would require more work on the caller's part as you described. I'm good to leave this as is for now.

File: relayer/message_coordinator.go [hunk 26]
```diff
@@ -252,3 +240,20 @@ func (mc *MessageCoordinator) ProcessBlock(blockHeader *types.Header, ethClient
 		go appRelayer.ProcessHeight(block.BlockNumber, handlers, errChan)
 	}
 }
+
+func FetchWarpMessageFromID(ethClient ethclient.Client, warpID common.Hash, blockNum *big.Int) (*relayerTypes.WarpMessageInfo, error) {
```
Comment (cam-schultz) [inline relayer/message_coordinator.go:244]: Naming nit: `FetchWarpMessage` seems sufficient to describe the function. Having the Warp ID in both the function name and as a parameter seems redundant.

File: README.md [hunk 27]
```diff
@@ -316,6 +316,54 @@ The relayer consists of the following components:
...
 </div>
 
+### API
+
+#### `/relay`
+- Used to manually relay a warp message. The body of the request must contain the following JSON:
+```json
+{
+ "blockchain-id": "<cb58-encoding of blockchain ID>",
+ "message-id": "<Hex encoding of warp message ID>",
```
[the remark this answers is not in the document]
Comment (cam-schultz) [reply README.md:326]: We should also use `ids.ID` for `warpID` variables and arguments internally.

File: README.md [hunk 28]
```diff
@@ -316,6 +316,54 @@ The relayer consists of the following components:
...
+ "block-num": <Integer representation of the block number that the message was sent in>
```
Comment (feuGeneA) [inline README.md:327]: I guess you left off the quotes around the value because it's an integer instead of a string, but the markdown render is in some weird highlight like its an error or something, pretty distracting.

File: README.md [hunk 29]
```diff
@@ -316,6 +316,54 @@ The relayer consists of the following components:
...
+#### `/health`
+- Takes no arguments. Returns a `200` status code if all Application Relayers are healthy. Returns a `503` status if any of the Application Relayers have experienced an unrecoverable error. Here is an example return body:
+```json
+{
+  "status": "down",
+  "details": {
+    "relayers-all": {
+      "status": "down",
+      "timestamp": "2024-06-01T05:06:07.685522Z",
+      "error": "<List of cb-58 encoded IDs for unhealth relayers>"
```
Comment (feuGeneA) [inline README.md:361]:
```suggestion
      "error": "<List of cb-58 encoded IDs for unhealthy relayers>"
```
what kind of list is this? not a json array? a CSV?

File: api/relay_message.go [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+			return
+		}
+
+		resp, err := json.Marshal(
+			RelayMessageResponse{
+				TransactionHash: txHash.Hex(),
+			},
+		)
+		if err != nil {
+			http.Error(w, "error writing response: "+err.Error(), http.StatusInternalServerError)
```
Comment (feuGeneA) [inline api/relay_message.go:81]:
```suggestion
			http.Error(w, "error marshaling response: "+err.Error(), http.StatusInternalServerError)
```

File: api/relay_message.go [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+			return
+		}
+
+		_, _ = w.Write(resp)
```
Comment (feuGeneA) [inline api/relay_message.go:85]: Is there a reason not to do error checking here?

File: api/relay_message.go [hunk 32]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (feuGeneA) [inline api/relay_message.go:126]: Is there a reason not to do error checking here?

File: api/relay_message.go [hunk 33]
```diff
@@ -0,0 +1,128 @@
...
+			return
+		}
+
+		resp, err := json.Marshal(
+			RelayMessageResponse{
+				TransactionHash: txHash.Hex(),
+			},
+		)
+		if err != nil {
+			http.Error(w, "error writing response: "+err.Error(), http.StatusInternalServerError)
```
Comment (feuGeneA) [inline api/relay_message.go:122]:
```suggestion
			http.Error(w, "error marshalling response: "+err.Error(), http.StatusInternalServerError)
```

File: api/relay_message.go (hunk 31 above)
Comment (geoff-vball) [reply api/relay_message.go:85]: So, if we error writing the response, there's really nothing we can tell the user, because we'd have to do so by writing something in the response... But I've plumbed the logger into our endpoint callers so we can log any errors here.

File: api/relay_message.go [hunk 34]
```diff
@@ -21,7 +22,7 @@ const (
 type RelayMessageRequest struct {
 	// Required. cb58 encoding of the source blockchain ID for the message
 	BlockchainID string `json:"blockchain-id"`
-	// Required. Hex encoding of the warp message ID
+	// Required. cb58 encoding of the warp message ID
 	MessageID string `json:"message-id"`
 	// Required. Integer representation of the block number that the message was sent in
 	BlockNum string `json:"block-num"`
```
Comment (cam-schultz) [inline api/relay_message.go:28]: Is there a reason we don't use an int here?

File: main/main.go [hunk 35]
```diff
@@ -202,8 +201,8 @@ func main() {
 
 	// Each Listener goroutine will have an atomic bool that it can set to false to indicate an unrecoverable error
 	api.HandleHealthCheck(relayerHealth)
```
Comment (cam-schultz) [inline main/main.go:203]: Can we pass the logger to the health check handler as well, and emit a fatal log when triggered? I think those logs would be useful to have directly from the application, and it would make that handler interface consistent with the other two.

**Review by cam-schultz — COMMENTED** [review]
Two more final comments, then I think we're good to go

File: api/relay_message.go (hunk 34 above)
Comment (geoff-vball) [reply api/relay_message.go]: Nope, changed :)

File: relayer/message_coordinator.go [hunk 36]
```diff
@@ -0,0 +1,259 @@
...
+// processes multiple messages (using their corresponding MessageHandlers) in a single shot.
+func (mc *MessageCoordinator) getAppRelayerMessageHandler(
+	warpMessageInfo *relayerTypes.WarpMessageInfo,
+) (
+	*ApplicationRelayer,
+	messages.MessageHandler,
+	error,
+) {
+	// Check that the warp message is from a supported message protocol contract address.
+	messageHandlerFactory, supportedMessageProtocol := mc.messageHandlerFactories[warpMessageInfo.UnsignedMessage.SourceChainID][warpMessageInfo.SourceAddress]
```
Comment (michaelkaplan13) [inline relayer/message_coordinator.go:62]: Nit: This line looks very long...I'm not too worried about it specifically but surprised we're not getting a linter error/warning in CI

**Review by michaelkaplan13 — COMMENTED** [review]
Last few nits, but LGTM otherwise

File: relayer/message_coordinator.go (hunk 36 above)
Comment (cam-schultz) [reply relayer/message_coordinator.go:62]: It looks like this can be enabled with the `lll` linter: https://golangci-lint.run/usage/linters/#lll

Comment (geoff-vball) [reply relayer/message_coordinator.go:62]: There are a good number of lines that don't fit in the line length, so I'll make a separate ticket for this.

**Review by cam-schultz — COMMENTED** [review]
One last uber mega nit that I'm just noticing now - can we capitalize the first letters of the log message strings for consistency? I think the only files that need to be modified are `health_check.go` and `relay_message.go`

**Review by cam-schultz — APPROVED** [review]
LGTM! 🙏

**Review by bernard-avalabs — APPROVED** [review]
LGTM
